libs/Microsoft.MixedReality.WebRTC/PeerConnection.cs

```diff
@@ -280,11 +280,201 @@ public struct VideoCaptureFormat
         public uint fourcc;
     }
 
+    /// <summary>
+    /// Settings to create a new transceiver wrapper.
+    /// </summary>
+    public class TransceiverInitSettings
+    {
+        /// <summary>
+        /// Transceiver name, for logging and debugging.
+        /// </summary>
+        public string Name;
+
+        /// <summary>
+        /// Initial value of <see cref="Transceiver.DesiredDirection"/>.
+        /// </summary>
+        public Transceiver.Direction InitialDesiredDirection = Transceiver.Direction.SendReceive;
+
+        /// <summary>
+        /// List of stream IDs to associate the transceiver with.
+        /// </summary>
+        public List<string> StreamIDs;
+    }
+
+    /// <summary>
+    /// Wrapper for an event possibly delayed.
+    /// </summary>
+    public class DelayedEvent
+    {
+        /// <summary>
+        /// The event handler.
+        /// </summary>
+        public Action Event;
+
+        /// <summary>
+        /// Begin suspending <see cref="Event"/>. This must be matched with a call
+        /// to <see cref="EndSuspend"/>. During this time, calling <see cref="Invoke"/>
+        /// does not invoke the event but instead queue it for later invoking by
+        /// <see cref="EndSuspend"/>.
+        /// </summary>
+        public void BeginSuspend()
+        {
+            lock (_lock)
+            {
+                ++_suspendCount;
+            }
+        }
+
+        /// <summary>
+        /// End suspending <see cref="Event"/> and invoke it if any call to <see cref="Invoke"/>
+        /// was made since the first <see cref="BeginSuspend"/> call.
+        /// </summary>
+        public void EndSuspend()
+        {
+            Action cb = null;
+            lock (_lock)
+            {
+                Debug.Assert(_suspendCount > 0);
+                --_suspendCount;
+                if (_eventPending && (_suspendCount == 0))
+                {
+                    cb = Event;
```

Comment:
`Event` is not set under lock. Unless we want to serialize changes to it together with calls `Begin`/`EndSuspend` - in which case we should set it under lock - there is no reason to read it under lock.

Reply:
It's not about reading it under the lock, it's about reading vs. not reading it, which tells whether it should be called now or not. And this decision depends on variables under the lock. I could instead read it unconditionally and have a boolean `shouldCallNow`, is that more clear?

Reply:
Yes, or just return early if no call is necessary.

Same of `Invoke()` below.

Reply:
Fixed.